### PR TITLE
filter_watermark: The initial commit for filter watermark

### PR DESCRIPTION
**Pull request by ginobiliwang**
Signed-off-by: Fenggang <ginobiliwang@gmail.com>

1) This is initial commit for watermark filter feature.
2) The idea is from a paper ([The Dataflow Model: A Practical Approach to Balancing Correctness, Latency, and Cost in Massive-Scale, Unbounded, Out-of-Order Data Processing](https://research.google/pubs/pub43864/)) from Google.
3) From stream processing perspective current function is limited. It has a fixed watermark. It can deal with out of order data. It has a batch window. It has tolerance for data latency.
4) Document PR for this filter watermark feature would be submitted soon.
5) The little demo and valgrind output below are for your reference.

+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

[root@localhost ~]# nc 127.0.0.1 5170
{"time":"2021-11-12 18:30:00"}
{"time":"2021-11-12 18:30:01"}
{"time":"2021-11-12 18:30:02"}
{"time":"2021-11-12 18:30:07"}
{"time":"2021-11-12 18:30:06"}
{"time":"2021-11-12 18:30:04"}
{"time":"2021-11-12 18:30:05"}
{"time":"2021-11-12 18:31:00"}

+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

[root@localhost build]# valgrind --leak-check=full  bin/fluent-bit -i tcp   -F watermark  -p 'time_field=time' 'watermark=20'    -m '*' -o stdout
==3805== Memcheck, a memory error detector
==3805== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==3805== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==3805== Command: bin/fluent-bit -i tcp -F watermark -p time_field=time watermark=20 -m * -o stdout
==3805==
Fluent Bit v1.7.4
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/05/08 22:12:14] [ info] [engine] started (pid=3805)
[2021/05/08 22:12:14] [ info] [storage] version=1.1.1, initializing...
[2021/05/08 22:12:14] [ info] [storage] in-memory
[2021/05/08 22:12:14] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/05/08 22:12:14] [ info] [input:tcp:tcp.0] listening on 0.0.0.0:5170
[2021/05/08 22:12:14] [ info] [sp] stream processor started
==3805== Warning: client switching stacks?  SP change: 0x9bc2788 --> 0xa48a850
==3805==          to suppress, use: --max-stackframe=9208008 or greater
==3805== Warning: client switching stacks?  SP change: 0xa48a7b8 --> 0x9bc2788
==3805==          to suppress, use: --max-stackframe=9207856 or greater
==3805== Warning: client switching stacks?  SP change: 0x9bc2788 --> 0xa48a7b8
==3805==          to suppress, use: --max-stackframe=9207856 or greater
==3805==          further instances of this message will not be shown.
[0] tcp.0: [1620483146.702023972, {"time"=>"2021-11-12 18:30:00"}]
[1] tcp.0: [1620483150.354546260, {"time"=>"2021-11-12 18:30:01"}]
[2] tcp.0: [1620483153.771547131, {"time"=>"2021-11-12 18:30:02"}]
[3] tcp.0: [1620483176.174986006, {"time"=>"2021-11-12 18:30:04"}]
[4] tcp.0: [1620483180.376079757, {"time"=>"2021-11-12 18:30:05"}]
[5] tcp.0: [1620483169.903685948, {"time"=>"2021-11-12 18:30:06"}]
[6] tcp.0: [1620483163.781975807, {"time"=>"2021-11-12 18:30:07"}]
^C[2021/05/08 22:13:19] [engine] caught signal (SIGINT)
[2021/05/08 22:13:19] [ warn] [engine] service will stop in 5 seconds
[2021/05/08 22:13:24] [ info] [engine] service stopped
==3805==
==3805== HEAP SUMMARY:
==3805==     in use at exit: 99,816 bytes in 3,380 blocks
==3805==   total heap usage: 4,229 allocs, 849 frees, 5,092,444 bytes allocated
==3805==
==3805== LEAK SUMMARY:
==3805==    definitely lost: 0 bytes in 0 blocks
==3805==    indirectly lost: 0 bytes in 0 blocks
==3805==      possibly lost: 0 bytes in 0 blocks
==3805==    still reachable: 99,816 bytes in 3,380 blocks
==3805==         suppressed: 0 bytes in 0 blocks
==3805== Reachable blocks (those to which a pointer was found) are not shown.
==3805== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==3805==
==3805== For lists of detected and suppressed errors, rerun with: -s
==3805== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
[root@localhost build]#

### CODE REVIEW

**Comment by ginobiliwang** [comment]
https://github.com/fluent/fluent-bit-docs/pull/530

File: plugins/CMakeLists.txt [hunk 1]
```diff
@@ -204,14 +204,14 @@ REGISTER_OUT_PLUGIN("out_websocket")
 REGISTER_OUT_PLUGIN("out_cloudwatch_logs")
 REGISTER_OUT_PLUGIN("out_kinesis_firehose")
 REGISTER_OUT_PLUGIN("out_kinesis_streams")
-REGISTER_OUT_PLUGIN("out_prometheus_exporter")
```
Comment (nigels-com) [inline plugins/CMakeLists.txt:207]: Intentional change?

**Comment by nigels-com** [comment]
Is there an well known and tested 3rd party heap implementation, as an alternative?

File: plugins/CMakeLists.txt (hunk 1 above)
Comment (ginobiliwang) [reply plugins/CMakeLists.txt:207]: Could be some merge issue, let me fix it.

**Comment by ginobiliwang** [comment]
> Is there an well known and tested 3rd party heap implementation, as an alternative?

Most of implementations are customized versions. For this plugin, i have done customization as well.

**Comment by ginobiliwang** [comment]
@edsiper @koleini , could you please review this PR ?

**Comment by edsiper** [comment]
@koleini ping

**Comment by ginobiliwang** [comment]
@koleini ,can you help to review this PR ?

**Comment by koleini** [comment]
This pull request requires documentation and unit tests to be included. In addition, there are indentation problems in the code that needs to be fixed.

Please see https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md for the guidelines.

**Comment by ginobiliwang** [comment]
Hi @koleini , this is the doc PR https://github.com/fluent/fluent-bit-docs/pull/530 .

Let me fix the format issue and fill up unit tests.

File: plugins/CMakeLists.txt (hunk 1 above)
Comment (nigels-com) [reply plugins/CMakeLists.txt:207]: Still looks wrong in the diff, does this need rebasing?

Comment (ginobiliwang) [reply plugins/CMakeLists.txt:207]: ok, i will rebase it in a whole after Masoud's initial review.

File: plugins/filter_watermark/watermark.c [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+    msgpack_object map;
+    msgpack_object key;
+    msgpack_object value;
+    int i; 
+    int map_size; 
+    int ret;
+    struct tm tm0;
+
+    msgpack_unpacked result;
+    msgpack_unpacked * destroy_result;
```
Comment (koleini) [inline plugins/filter_watermark/watermark.c:231]: `*destroy_result`

File: plugins/filter_watermark/watermark.c [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+        c_heap_insert(h, record_ptr);
+     }
+
+     /* main loop to flush out data */
+     while(tl - watermark >= win_right_edge) {
+         /*Dump all data into array to flush out*/
+         tmp_record = c_heap_read_root(h);
+         if (tmp_record == NULL) {
+             /* This place could not be reached, since at least one piece of record is in the heap*/
+             flb_plg_error(context->ins,  "Fatal Error!");
```
Comment (koleini) [inline plugins/filter_watermark/watermark.c:149]: Extra space?

File: plugins/filter_watermark/watermark.c [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+                    flb_free(tmp_ptr);
+                }
+                ctx->record_count = 0;
+	    } else if (ret == -1) {
+		/* record is too late or some other fatal error, we just clean up memory*/
+                msgpack_sbuffer_free(tmp_sbuf);
+                flb_free(tmp_ptr);
+            } else {
+                flb_plg_debug(ctx->ins, "No record pop up in this round");
+                //return FLB_FILTER_MODIFIED;
```
Comment (koleini) [inline plugins/filter_watermark/watermark.c:323]: Remove

**Comment by koleini** [comment]
Would be good to add tests if possible.

**Comment by ginobiliwang** [comment]
> Would be good to add tests if possible.

Sorry，please ignore the request review, i clicked it by accident.
I have tried to add some test cases, but failed to make it, will ping you separately for it.

File: plugins/filter_watermark/watermark.c (hunk 3 above)
Comment (ginobiliwang) [reply plugins/filter_watermark/watermark.c]: fixed in commit https://github.com/fluent/fluent-bit/commit/4b46f5c1048639ed37cafc345e575f6794dccba5

File: plugins/filter_watermark/watermark.c (hunk 2 above)
Comment (ginobiliwang) [reply plugins/filter_watermark/watermark.c]: fixed in commit https://github.com/fluent/fluent-bit/commit/4b46f5c1048639ed37cafc345e575f6794dccba5

**Comment by ginobiliwang** [comment]
@koleini  please review latest code if possible.

File: plugins/filter_watermark/heap.c [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    left = (2 * root) + 1;
+    if (left >= h->array_len)
+        left = 0;
+
+    right = (2 * root) + 2;
+    if (right >= h->array_len)
+        right = 0;
+
+    /* Check which one of the children is smaller. */
+    if ((left == 0) && (right == 0))
```
Comment (koleini) [inline plugins/filter_watermark/heap.c:55]: Still some of your conditional statements don't follow the guidelines for brace usage (please find and fix the rest).

File: plugins/filter_watermark/heap.c [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    }
+}
+
+
+
+struct c_heap_t *c_heap_create(int (*compare)(void *, void *), int (*deconstructor)(void *)) 
+{
+    struct c_heap_t *h;
+
+    if (compare == NULL)
```
Comment (koleini) [inline plugins/filter_watermark/heap.c:98]: Requires braces (please find the rest and fix).

File: plugins/filter_watermark/heap.c [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        return NULL;
+
+    if (deconstructor == NULL)
+        return NULL;
+
+    h = calloc(1, sizeof(*h));
```
Comment (koleini) [inline plugins/filter_watermark/heap.c:104]: Please use `flb_calloc` instead.

File: plugins/filter_watermark/heap.c [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+    size_t index;
+    void **tmp;
+
+    if ((h == NULL) || (ptr == NULL))
+        return -EINVAL;
+
+    assert(h->array_len <= h->array_size);
+    if (h->array_len == h->array_size) {
+
+        tmp = realloc(h->array, (h->array_size + 16) * sizeof(*h->array));
```
Comment (koleini) [inline plugins/filter_watermark/heap.c:146]: Please use `flb_realloc`.

File: plugins/filter_watermark/heap.c [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+        h->array[h->array_len - 1] = NULL;
+        h->array_len--;
+
+        reheap(h, /* root = */ 0, DIR_DOWN);
+    }
+
+    /* free some memory */
+    if ((h->array_len + 32) < h->array_size) {
+
+        tmp = realloc(h->array, (h->array_len + 16) * sizeof(*h->array));
```
Comment (koleini) [inline plugins/filter_watermark/heap.c:193]: Please use `flb_realloc`.

File: plugins/filter_watermark/heap.c [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+    for(i=0; i< h->array_len; i++) {
+        h->deconstructor(h->array[i]);
+    }
+
+    h->array_len = 0;
+    h->array_size = 0;
+    free(h->array);
+    h->array = NULL;
+
+    free(h);
```
Comment (koleini) [inline plugins/filter_watermark/heap.c:132]: Please use `flb_free` (and the same for other places).

File: plugins/filter_watermark/watermark.c [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+    
+    if (!str) {
+        flb_plg_error(ctx->ins, "time_field is not defined!");
+        return -1;
+    }
+    ctx->time_field = flb_sds_create(str);
+    return 0;
+}
+
+int deconstructor_imp(void *v) {
```
Comment (koleini) [inline plugins/filter_watermark/watermark.c:80]: Looks like indentation is not right here and some other places. You can use some C syntax beautifiers to identify/fix.

File: plugins/filter_watermark/watermark.c [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+            /* we reply on strptime to check the format. */
+            memset(&tm0, 0, sizeof(struct tm));
+            /* strptime("2021-11-12 18:31:01", "%Y-%m-%d %H:%M:%S", &tm0); */
+            if( strptime(tmp_buf, "%Y-%m-%d %H:%M:%S", &tm0) == 0) {
+                flb_plg_info(ctx->ins, "Timestamp format is not acceptable by strptime, drop the record\n");
+		continue;
+            }
+
+           /* when it reaches here, we need to malloc memory to store all the content of record and insert record into heap. */
+	   struct flb_filter_watermark_record *tmp_ptr = flb_malloc(sizeof(struct flb_filter_watermark_record));
```
Comment (koleini) [inline plugins/filter_watermark/watermark.c:297]: Variables must be declared at the beginning of a function and not in the middle of the code (see guidelines).

**Comment by koleini** [comment]
In addition to the fixes, a proper documentation on this plugin is needed. For instance, users need to know the definition of `batch window` and how it works for watermarking.

**Comment by ginobiliwang** [comment]
will continue to finish this pr soon, tied up by some other affairs.

**Comment by ginobiliwang** [comment]
Hi @leonardo-albertovich, sure. Thanks so much for comments.

For data structure, it is just a simple `heap` structure.  I do not think binary tree can replace it, but if possible could you share with me the binary tree implementation ? i could take a reference.

Thanks for information about "to use the log event abstraction layer", i will ping you on slack for details.

**Comment by leonardo-albertovich** [comment]
The binary tree we use is this well known red/black binary tree implementation https://github.com/fluent/fluent-bit/tree/master/lib/rbtree

**Comment by ginobiliwang** [comment]
I did not notice we have rbtree in place to use. And definitely it is ok to use rbtree for current feature.

**Comment by leonardo-albertovich** [comment]
I'm glad you find it fitting, I agree and would really appreciate it if at some point you updated your code to use it.

Thanks a lot!

**Comment by ginobiliwang** [comment]
will continue to finish this pr soon, tied up by some other affairs.

**Comment by ginobiliwang** [comment]
sorry for the delay, will catch up ASAP.

**Comment by ginobiliwang** [comment]
> I'm glad you find it fitting, I agree and would really appreciate it if at some point you updated your code to use it.
> 
> Thanks a lot!

**Comment by ginobiliwang** [comment]
HI @leonardo-albertovich , i have tried to use [rbtree in fluent bit repo]( https://github.com/fluent/fluent-bit/tree/master/lib/rbtree) to failed to apply. The main issue is that rbtree could not handle duplicate key.

**Comment by ginobiliwang** [comment]
Hi @leonardo-albertovich , @koleini , @edsiper  any chance to reopen this PR ? i can not re-open it by myself.

**Comment by lecaros** [comment]
ping @ginobiliwang

**Comment by ginobiliwang** [comment]
> ping @ginobiliwang

Thanks @lecaros for reopening the PR. The same question i have post last year, re-paste here.
=======>
i have tried to use [rbtree in fluent bit repo](https://github.com/fluent/fluent-bit/tree/master/lib/rbtree) , but failed to apply. The main issue is that rbtree could not handle duplicate key. In this PR, it would need to deal with records with duplicate keys. So i have to revert back to use heap data structure to implement it, i think heap is the best data structure to do the job.
=======>

Please review the above, thanks.

**Comment by lecaros** [comment]
pinging @edsiper @leonardo-albertovich and @cosmo0920  to help you out with your question

**Comment by ginobiliwang** [comment]
pinging @edsiper @leonardo-albertovich and @cosmo0920 to help you out with your question

**Pull request closed without merging**